docs/eBPFWeakReferences.md

```diff
@@ -0,0 +1,82 @@
+
+**Introduction and problem statement**
+
+The eBPF user-kernel interface is built around a few key objects such as 'Programs', 'Maps' and their interaction. The 'eBPF for Windows' implementation makes extensive use of 'reference counting' to manage the interactions and lifetimes of these objects.  Code paths 'acquire' references to the required objects for the needed duration and 'release' them when done.  The reference counting design ensures that the referred-to objects remain valid (i.e. are not deleted) for the duration of the reference acquisition.
+
+By default, these references are of the 'strong' type, i.e. the referred-to objects are guaranteed to be valid for the duration of their strong references.  While this property is mandatory in most use cases, it can lead to 'circular reference' instances.  Such instances are perfectly valid and the relationship between a 'Program Array Map' and a eBPF 'program' is an example of such a relationship.  This circular reference ensures that both the 'Program Array Map' and the eBPF 'program' remain valid for the duration of each other's existence, as they should be.
```

Comment:
```suggestion
By default, these references are of the 'strong' type, i.e., the referred-to objects are guaranteed to be valid for the duration of their strong references.  While this property is mandatory in most use cases, it can lead to 'circular reference' instances.  Such instances are perfectly valid and the relationship between a 'Program Array Map' and an eBPF 'program' is an example of such a relationship.  This circular reference ensures that both the 'Program Array Map' and the eBPF 'program' remain valid for the duration of each other's existence, as they should be.
```
I'm not sure if there's a legitimate case for a circular reference existing. (That doesn't mean the design described isn't a good one though.)  I think the real issue this design addresses is that it's tricky in general to prevent circular references given there can be a long chain of map1->prog1->map2->prog2->...->mapN->progN->map1 for example.

Reply:
While we don't _have_ circular reference instances currently, the intent was to solve the 'circular reference' problem in a _generic_ manner, be it direct (as shown by the simple example in this document) or the eminently possible chain that you'd illustrated.